src/python/analyzer_executor/tests/test_cache_integration.py

```diff
@@ -36,54 +46,76 @@ def _AnalyzerExecutorSingleton(stub_env=False, env_addr=None, env_port=None):
 
 
 @pytest.mark.integration_test
-def test_connection_info(AnalyzerExecutorSingleton) -> None:
+def test_connection_info(executor_fixture: ReturnsAnalyzerExecutor) -> None:
     """
     Ensures exceptions are raised for incomplete connection info.
     """
 
     with pytest.raises(ValueError):
-        ae = AnalyzerExecutorSingleton(
-            stub_env=True, env_addr=SAMPLE_ADDR, env_port=None
-        )
+        ae = executor_fixture(stub_env=True, env_addr=SAMPLE_ADDR, env_port=None)
 
     with pytest.raises(ValueError):
-        ae = AnalyzerExecutorSingleton(
-            stub_env=True, env_addr=None, env_port=SAMPLE_PORT
-        )
+        ae = executor_fixture(stub_env=True, env_addr=None, env_port=SAMPLE_PORT)
 
     with pytest.raises(ValueError):
-        ae = AnalyzerExecutorSingleton(stub_env=True, env_addr=None, env_port=None)
-
-
+        ae = executor_fixture(stub_env=True, env_addr=None, env_port=None)
+
+
+@hypothesis.given(
+    k1=NonemptyStringStrategy,
+    k2=NonemptyStringStrategy,
+)
+@hypothesis.settings(
+    # Doesn't like the Pytest fixture mixed with Hypothesis givens.
+    # It's okay, since the fixture just returns a function.
+    suppress_health_check=[hypothesis.HealthCheck.function_scoped_fixture]
+)
 @pytest.mark.integration_test
-def test_hit_cache(AnalyzerExecutorSingleton) -> None:
+def test_hit_cache(
+    executor_fixture: ReturnsAnalyzerExecutor,
+    k1: str,
+    k2: str,
+) -> None:
     """
     Initializes the AnalyzerExecutor singleton with Redis connection params
     sourced from the environment, expecting hit cache to populate.
     """
-    ae = AnalyzerExecutorSingleton(stub_env=False)
-
-    k1, k2 = st.text(min_size=3, max_size=64), st.text(min_size=3, max_size=64)
+    ae = executor_fixture(stub_env=False)
 
     assert not ae.check_hit_cache(k1, k2)
     ae.update_hit_cache(k1, k2)
     assert ae.check_hit_cache(k1, k2)
+    # Clean up, because Hypothesis provides duplicate inputs
+    ae.delete_hit_cache(k1, k2)
```

Comment:
rather than introducing a function `delete_hit_cache()` into the analyzer-executor's public API that is only used in a test, would it make more sense to just generate hypothesis data that's not already in the cache? This can be accomplished pretty straightforwardly with a [`data()` strategy](https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests) or by using [`assume()` within a composite strategy](https://hypothesis.readthedocs.io/en/latest/data.html#composite-strategies) or by using the [`.filter()` combinator](https://hypothesis.readthedocs.io/en/latest/data.html#filtering)

Reply:
A more robust method might be to isolate the cache entirely--i.e. arrange the code such that each test execution gets a completely new cache. This might be accomplished using monkeypatching and the `@pytest.mark.forked` decorator. See e.g.:
https://github.com/grapl-security/grapl/blob/internal-78-analyzer-deployer-api/src/python/analyzer-deployer/tests/test_app.py#L66-L75
and
https://github.com/grapl-security/grapl/blob/internal-78-analyzer-deployer-api/src/python/analyzer-deployer/tests/test_app.py#L318-L361
however, I'm not 100% certain this will play nice with hypothesis...

Reply:
RE: the more-robust method, this is actually an integration test against the Redis service, so it has to be a real cache. No monkeypatching would work. 

RE: making a "only generate not-yet-used strings" strategy,
- that's definitely doable
- afaict it's considered non-hypothesis-ic
- quite frankly it's more effort than I want to spend on a random bug I found by accident :)

Reply:
The `analyzer-deployer` code I linked is also an integration test against real dynamodb tables (albeit a local one). The way it works is we have a function `app._analyzers_table()` which returns a `dynamdb.Table`. We monkeypatch that function in the test code to instead return a different table for each test. That way each test gets its own table and starts with fresh, uncorrupted state. I suspect we could devise a similar approach for our redis caches.

I believe you are correct about the hypothesis strategy stuff I proposed. I made a minimal example e.g.:

```python
@hypothesis.given(
    st.text(min_size=3, max_size=64).filter(lambda k: k not in DB),
    st.text(min_size=3, max_size=64),
)
def test_db(k, v):
    assert query_db(k) is None
    insert_into_db(k, v)
    assert query_db(k) == v
```

and got the following error:

```
E       hypothesis.errors.Flaky: Inconsistent data generation! Data generation behaved differently between different runs. Is your data generation depending on external state?
```

so I suspect we'd actually have to build a [stateful strategy](https://hypothesis.readthedocs.io/en/latest/stateful.html) here.